src/progress.rs

```diff
@@ -841,4 +1149,101 @@ mod test_progress_set {
         assert_eq!(pre, *set.get(1).expect("Peer should not have been deleted"));
         Ok(())
     }
+
+    #[test]
+    fn test_membership_change_configuration_remove_voter() -> Result<()> {
+        check_membership_change_configuration((vec![1, 2], vec![]), (vec![1], vec![]))
+    }
+
+    #[test]
+    fn test_membership_change_configuration_remove_learner() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![2]), (vec![1], vec![]))
+    }
+
+    #[test]
+    fn test_membership_change_configuration_conflicting_sets() {
+        assert!(
+            check_membership_change_configuration((vec![1], vec![]), (vec![1], vec![1]),).is_err()
+        )
+    }
+
+    #[test]
+    fn test_membership_change_configuration_empty_sets() {
+        assert!(check_membership_change_configuration((vec![], vec![]), (vec![], vec![])).is_err())
+    }
+
+    #[test]
+    fn test_membership_change_configuration_empty_voters() {
+        assert!(
+            check_membership_change_configuration((vec![1], vec![]), (vec![], vec![]),).is_err()
+        )
+    }
+
+    #[test]
+    fn test_membership_change_configuration_add_voter() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![]), (vec![1, 2], vec![]))
+    }
+
+    #[test]
+    fn test_membership_change_configuration_add_learner() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![]), (vec![1], vec![2]))
+    }
+
+    #[test]
+    fn test_membership_change_configuration_promote_learner() -> Result<()> {
+        check_membership_change_configuration((vec![1], vec![2]), (vec![1, 2], vec![]))
+    }
+
+    fn check_membership_change_configuration(
+        start: (impl IntoIterator<Item = u64>, impl IntoIterator<Item = u64>),
+        end: (impl IntoIterator<Item = u64>, impl IntoIterator<Item = u64>),
+    ) -> Result<()> {
+        let start_voters = start.0.into_iter().collect::<FxHashSet<u64>>();
+        let start_learners = start.1.into_iter().collect::<FxHashSet<u64>>();
+        let end_voters = end.0.into_iter().collect::<FxHashSet<u64>>();
+        let end_learners = end.1.into_iter().collect::<FxHashSet<u64>>();
+        let transition_voters = start_voters
+            .union(&end_voters)
+            .cloned()
+            .collect::<FxHashSet<u64>>();
+        let transition_learners = start_learners
+            .union(&end_learners)
+            .cloned()
+            .collect::<FxHashSet<u64>>();
+
+        let mut set = ProgressSet::default();
+        let default_progress = Progress::new(0, 10);
+
+        for starter in start_voters {
+            set.insert_voter(starter, default_progress.clone())?;
+        }
+        for starter in start_learners {
+            set.insert_learner(starter, default_progress.clone())?;
+        }
```

Comment:
Why is the initial `ProgressSet` created by calling `default` and then mutating it with accessors, when doing a membership change means passing in a complete `Configuration`? The latter seems more foolproof. The mutating accessors seem easy to misuse.

Thinking of creating the initial set itself as a membership change _sounds_ like a reasonable generalization.

Reply:
I'm confused by this comment. :( Maybe you can clarify it with me.